skills/dff_knowledge_prompted_skill/scenario/response.py

```diff
@@ -0,0 +1,173 @@
+GENERATIVE_SERVICE_CONFIG = getenv("GENERATIVE_SERVICE_CONFIG")
+USE_KG_DATA = int(getenv("USE_KG_DATA", 0))
+USER_KG_SERVICE_URL = getenv("USER_KG_SERVICE_URL")
+if GENERATIVE_SERVICE_CONFIG:
+    with open(f"common/generative_configs/{GENERATIVE_SERVICE_CONFIG}", "r") as f:
+        GENERATIVE_SERVICE_CONFIG = json.load(f)
+
+PROMPT_FILE = getenv("PROMPT_FILE")
+N_UTTERANCES_CONTEXT = int(getenv("N_UTTERANCES_CONTEXT", 3))
+ENVVARS_TO_SEND = getenv("ENVVARS_TO_SEND", None)
+ENVVARS_TO_SEND = [] if ENVVARS_TO_SEND is None else ENVVARS_TO_SEND.split(",")
+
+assert GENERATIVE_SERVICE_URL
+assert PROMPT_FILE
+assert USER_KG_SERVICE_URL
+
+with open(PROMPT_FILE, "r") as f:
+    PROMPT_DICT = json.load(f)
+PROMPT = PROMPT_DICT["prompt"]
+
+FIX_PUNCTUATION = re.compile(r"\s(?=[\.,:;])")
+PROMPT_REPLACEMENT_COMMAND = re.compile(r"^/prompt")
+PROMPT_RESET_COMMAND = re.compile(r"^/resetprompt")
+DEFAULT_CONFIDENCE = 0.9
+SUPER_CONFIDENCE = 1.0
+LOW_CONFIDENCE = 0.7
+
+
+def compose_data_for_model(ctx, actor):
+    # consider N_UTTERANCES_CONTEXT last utterances
+    context = int_ctx.get_utterances(ctx, actor)[-N_UTTERANCES_CONTEXT:]
+    context = [uttr.get("text", "") for uttr in context]
+
+    if context:
+        context = [re.sub(FIX_PUNCTUATION, "", x) for x in context]
+
+    return context
+
+
+def generative_response(ctx: Context, actor: Actor, *args, **kwargs) -> Any:
+    curr_responses, curr_confidences, curr_human_attrs, curr_bot_attrs, curr_attrs = (
+        [],
+        [],
+        [],
+        [],
+        [],
+    )
+
+    def gathering_responses(reply, confidence, human_attr, bot_attr, attr):
+        nonlocal curr_responses, curr_confidences, curr_human_attrs, curr_bot_attrs, curr_attrs
+        if reply and confidence:
+            curr_responses += [reply]
+            curr_confidences += [confidence]
+            curr_human_attrs += [human_attr]
+            curr_bot_attrs += [bot_attr]
+            curr_attrs += [attr]
+
+    dialog_context = compose_data_for_model(ctx, actor)
+    # get variables which names are in `ENVVARS_TO_SEND` (splitted by comma if many)
+    # from user_utterance attributes or from environment
+    human_uttr_attributes = int_ctx.get_last_human_utterance(ctx, actor).get("attributes", {})
+    lm_service_kwargs = human_uttr_attributes.pop("lm_service_kwargs", None)
+    lm_service_kwargs = {} if lm_service_kwargs is None else lm_service_kwargs
+    envvars_to_send = ENVVARS_TO_SEND if len(ENVVARS_TO_SEND) else human_uttr_attributes.get("envvars_to_send", [])
+    sending_variables = compose_sending_variables(
+        lm_service_kwargs,
+        envvars_to_send,
+        **human_uttr_attributes,
+    )
+
+    shared_memory = int_ctx.get_shared_memory(ctx, actor)
+    prompt = shared_memory.get("prompt", "")
+    prompt = PROMPT
+
+    custom_el = (
+        ctx.misc.get("agent", {})
+        .get("dialog", {})
+        .get("human_utterances", [{}])[-1]
```

Comment:
btw there is a special function `int_ctx.get_last_human_utterance(ctx, actor)` for that purpose. It has a check for validation, so better to use this function

Reply:
Done